File: 프로그래머스/1/178871. 달리기 경주/달리기 경주.java
import java.util.*;

class Solution {
    public String[] solution(String[] players, String[] callings) {
        Map<String, Integer> ranking = new HashMap<>();
        for(int i = 0; i < players.length; i++){
            ranking.put(players[i], i);
        }
        for(String player : callings){
            int playerIdx = ranking.get(player);
            
            String frontPlayer = players[playerIdx - 1];
            
            ranking.replace(frontPlayer, playerIdx);
            players[playerIdx] = frontPlayer;
            
            ranking.replace(player, playerIdx - 1);
            players[playerIdx - 1] = player; 
        }
        return players;
    }
}

File: 프로그래머스/1/178871. 달리기 경주/달리기 경주.swift
import Foundation

func solution(_ players:[String], _ callings:[String]) -> [String] {
    var ranking = Array<String>()
    var nameKey = Dictionary<String, Int>()
    var idxKey = Dictionary<Int, String>()
    for (idx, name) in players.enumerated() {
        nameKey[name] = idx
        idxKey[idx] = name
    }
    for call in callings {
        var idx = nameKey[call]!
        var previousIdx = idx - 1
        var previousName = idxKey[previousIdx]!
        idxKey[previousIdx] = call
        idxKey[idx] = previousName
        nameKey[call] = previousIdx
        nameKey[previousName] = idx
    }
    let count = players.count
    for i in 0..<count {
        var name = idxKey[i]!
        ranking.append(name)
    }
    return ranking
}